### DISCUSSION TITLE
Refresh token rotation and multi-tabs : is KeyCloak's behaviour the best one (for ex compared to Auth0) ?

### DISCUSSION THREAD

**Discussion by semangard** [keycloak/keycloak#10937] [Q&A]
Hello,

We tried to activate KeyCloak refresh token rotation:
![image](https://user-images.githubusercontent.com/57065741/160080076-776eb62c-b43f-4fcf-ae25-6bc6299afe86.png)

But as indicated by the diagram we are facing an "issue" with a multi-tabs SPA (same user => same SSO session) : all the RT of a SSO session are revoked even if no reuse has been detected (so not only in a theft case): tab A and tab B have different RT and the use of RTA1 by tab A (RTA1 has not been used yet) triggers the revocation of RTB2 of tab B.

According to doc of other IdP like Auth0 the implemtation is slightly different: the revocation of all the RT of a SSO session happens only when a RT is reused (so only in a theft case) => it is called "_Refresh Token Rotation **with Reuse Detection**_"
https://auth0.com/blog/securing-single-page-applications-with-refresh-token-rotation/

![image](https://user-images.githubusercontent.com/57065741/160087396-e55cc181-2b73-412d-8518-64777169bf68.png)

Note also that Auth0 defines a _reuse interval_ and not a _max reuse_ (nb)

So, what is the best behaviour ? And sould KeyCloak change its behaviour ?

Thanks and regards

**Comment by jamesbruice** [comment]
the reuse detection kicks in when the second tab is trying to refresh the token

as a result all refresh_tokens are being revoked

**Comment by juluanb** [comment]
We encountered the same problem. Is the deactivation of the "Revoke Refresh Token" the only solution to make that works as expected? Are there potential security issues doing that?

**Reply by Badisi** [reply]
The answer is "yes" and "yes", unfortunately.

Discussion on the topic can be found here : https://github.com/keycloak/keycloak/issues/14122
But current status is : "there is currently no capacity within the team to address this issue"
